core/src/main/scala/org/apache/spark/scheduler/OutputCommitCoordinator.scala

```diff
@@ -0,0 +1,172 @@
+
+  private val authorizedCommittersByStage:
+  mutable.Map[StageId, mutable.Map[TaskId, TaskAttemptId]] = mutable.HashMap()
+
+  def stageStart(stage: StageId) {
+    coordinatorActor ! StageStarted(stage)
+  }
+  def stageEnd(stage: StageId) {
+    coordinatorActor ! StageEnded(stage)
+  }
+
+  def canCommit(
+      stage: StageId,
+      task: TaskId,
+      attempt: TaskAttemptId): Boolean = {
+    AkkaUtils.askWithReply(AskPermissionToCommitOutput(stage, task, attempt),
+      coordinatorActor, maxAttempts, retryInterval, timeout)
+  }
+
+  def taskCompleted(
+      stage: StageId,
+      task: TaskId,
+      attempt: TaskAttemptId,
+      successful: Boolean) {
+    coordinatorActor ! TaskCompleted(stage, task, attempt, successful)
+  }
+
+  def stop() {
+    val stopped = AkkaUtils.askWithReply[Boolean](StopCoordinator, coordinatorActor, timeout)
+    if (!stopped) {
+      logWarning("Expected true from stopping output coordinator actor, but got false!")
+    }
+    authorizedCommittersByStage.foreach(_._2.clear)
+    authorizedCommittersByStage.clear
+  }
+
+  private def handleStageStart(stage: StageId): Unit = {
+    authorizedCommittersByStage(stage) = mutable.HashMap[TaskId, TaskAttemptId]()
+  }
+
+  private def handleStageEnd(stage: StageId): Unit = {
+    authorizedCommittersByStage.remove(stage)
+  }
+
+  private def handleAskPermissionToCommit(
+      stage: StageId,
+      task: TaskId,
+      attempt: TaskAttemptId):
+      Boolean = {
+    if (!authorizedCommittersByStage.contains(stage)) {
+      logDebug(s"Stage $stage has completed, so not allowing task attempt $attempt to commit")
+      return false
+    }
+    val authorizedCommitters = authorizedCommittersByStage(stage)
+    if (authorizedCommitters.contains(task)) {
+      val existingCommitter = authorizedCommitters(task)
+      logDebug(s"Denying $attempt to commit for stage=$stage, task=$task; " +
+        s"existingCommitter = $existingCommitter")
+      false
+    } else {
+      logDebug(s"Authorizing $attempt to commit for stage=$stage, task=$task")
+      authorizedCommitters(task) = attempt
+      true
+    }
+  }
+
+  private def handleTaskCompletion(
+      stage: StageId,
+      task: TaskId,
+      attempt: TaskAttemptId,
+      successful: Boolean): Unit = {
+    if (!authorizedCommittersByStage.contains(stage)) {
+      logDebug(s"Ignoring task completion for completed stage")
+      return
+    }
+    val authorizedCommitters = authorizedCommittersByStage(stage)
+    if (authorizedCommitters.get(task) == Some(attempt) && !successful) {
+      logDebug(s"Authorized committer $attempt (stage=$stage, task=$task) failed; clearing lock")
+      // The authorized committer failed; clear the lock so future attempts can commit their output
+      authorizedCommitters.remove(task)
+    }
+  }
+
+}
+
+private[spark] object OutputCommitCoordinator {
+
+  class OutputCommitCoordinatorActor(outputCommitCoordinator: OutputCommitCoordinator)
```

Comment:
I made all of the handleX methods on OutputCommitCoordinator private, so that the protocol enforced is to call the methods that send the message to the actor. The actor could just be built using a constructor, I agree - some misunderstanding on my end on how Scala works in SparkEnv.registerOrLookup, and will fix that.

However since @vanzin described an alternative approach to using the actor, this may change soon. Will be working on this today.